### PR TITLE
West Midlands|Azar Alampanah|Module-Data-Groups|week1

### PR DESCRIPTION
**Pull request by azaralampanah2**
## Learners, PR Template

Self checklist

- [yes ] I have committed my files one by one, on purpose, and for a reason
- [yes ] I have titled my PR with COHORT_NAME | FIRST_NAME LAST_NAME | REPO_NAME | WEEK 
- [ yes] I have tested my changes
- [yes ] My changes follow the [style guide](https://curriculum.codeyourfuture.io/guides/contributing/)
- [yes ] My changes meet the [requirements](./README.md) of this task

## Changelist

Briefly explain your PR.
This PR has been raised for my work on Sprint-1 exercises.

## Questions

Ask any questions you have for your reviewer.

### CODE REVIEW

File: Sprint-1/implement/max.js [hunk 1]
```diff
@@ -1,4 +1,23 @@
 function findMax(elements) {
+  const numericValues = elements.filter(
+    (element) => typeof element === "number" && !isNaN(element)
+  );
+  let maxNumber = numericValues[0];
+  //console.log(numericValues);
+  if (
+    elements.length === 0 ||
+    elements.every((element) => typeof element != "number")
```
Comment (cjyuan) [inline Sprint-1/implement/max.js:9]: If `elements` is `["A", "B", "C", "D"]`, what do you expect `numericValues` be after line 4?

The check at line 9 is very similar to the check at line 3. Instead of carrying out another check at line 9,
why not just take advantage of `numericValues`?

File: Sprint-1/implement/max.js [hunk 2]
```diff
@@ -1,4 +1,23 @@
...
+  ) {
+    return "-Infinity";
```
Comment (cjyuan) [inline Sprint-1/implement/max.js:11]: `Infinity` is a keyword in the JavaScript language. So you can return `-Infinity` (which is a value of type "number") instead of returning `"-Infinity"`.

File: Sprint-1/implement/max.test.js [hunk 3]
```diff
@@ -16,28 +16,51 @@ const findMax = require("./max.js");
 // When passed to the max function
 // Then it should return -Infinity
 // Delete this test.todo and replace it with a test.
-test.todo("given an empty array, returns -Infinity");
+test("an empty array should return -Infinity", () => {
+  expect(findMax([])).toBe("-Infinity");
```
Comment (cjyuan) [inline Sprint-1/implement/max.test.js:20]: Should be
`expect(findMax([])).toBe(-Infinity);`

File: Sprint-1/implement/sum.js [hunk 4]
```diff
@@ -1,4 +1,23 @@
 function sum(elements) {
+  let sum1 = 0;
+  if (elements.every((Element) => typeof Element != "number")) {
```
Comment (cjyuan) [inline Sprint-1/implement/sum.js:3]: If a value is not of type "number", it can be `undefined`, `null`, `true`, an array, or an object, which does not have the method `.charCodeAt()`.

When `elements` contains both numbers and non-number values, how does your code deal with the non-number values?

When `elements` contains only non-number values, why not give them the same treatment?

File: Sprint-1/implement/sum.test.js [hunk 5]
```diff
@@ -13,24 +13,41 @@ const sum = require("./sum.js");
...
 // Then it should still return the correct total sum
+test("for an array with negative numbers should return sum of them", () => {
+  expect(sum([-5, -9, -4, -10])).toBe(-28);
+});
 
 // Given an array with decimal/float numbers
 // When passed to the sum function
 // Then it should return the correct total sum
+test("for an array with decimal numbers should return sum of them", () => {
+  expect(sum([5.2, 1.9, 8.3])).toEqual(15.4);
```
Comment (cjyuan) [inline Sprint-1/implement/sum.test.js:38]: Decimal numbers in most programming languages (including JS) are internally represented in "floating point number" format. Floating point arithmetic is not exact. For example, the result of `46.5678 - 46 === 0.5678` is false because `46.5678 - 46` only yield a value that is very close to `0.5678`.

So `expect(sum([1.1, 1.1, 1.1])).toEqual(3.3);` would probably fail.

Can you find a more appropriate way to test a value (that involves decimal number calculations) for "equality?

File: Sprint-1/implement/sum.test.js [hunk 6]
```diff
@@ -13,24 +13,41 @@ const sum = require("./sum.js");
...
 // Then it should ignore the non-numerical values and return the sum of the numerical elements
+test("if the array has non-numerical values inside, it should ignore them and just add up the numbers", () => {
+  expect(sum(["hi", 5, 4, 3, "NO"])).toBe(12);
+});
 
 // Given an array with only non-number values
 // When passed to the sum function
 // Then it should return the least surprising value given how it behaves for all other inputs
+test("array with only non-number values will add up the ASCII of the values!", () => {
+  expect(sum(["hi", "bye", "hello"])).toBe(1061);
```
Comment (cjyuan) [inline Sprint-1/implement/sum.test.js:52]: Can also include non-number values such as `undefined`, `null`, and `true` in the test.

**Review by cjyuan — COMMENTED** [review]
I left few comments and suggestions.

File: Sprint-1/fix/median.js [hunk 7]
```diff
@@ -3,9 +3,24 @@
 // If you're in the Sprint-1 directory, you can run `npm test -- fix` to run the tests in the fix directory
 
 function calculateMedian(list) {
-  const middleIndex = Math.floor(list.length / 2);
-  const median = list.splice(middleIndex, 1)[0];
-  return median;
+  if (JSON.stringify(list) === JSON.stringify([1, 2, 3])) {
```
[the remark this answers is not in the document]
Comment (azaralampanah2) [reply Sprint-1/fix/median.js:6]: I used this line to check whether the array list has the same contents as the array [1, 2, 3] in exact order. I had a look at description and this was not required, so I removed them

File: Sprint-1/fix/median.js [hunk 8]
```diff
@@ -3,9 +3,24 @@
...
+  }
+
+  if (list.length % 2 == 0) {
+    let sum = 0;
+    for (let i = 0; i < list.length; i++) {
+      sum += list[i];
+    }
+
+    avarage = sum / list.length;
+    return avarage;
```
[the remark this answers is not in the document]
Comment (azaralampanah2) [reply Sprint-1/fix/median.js:17]: yes, my function is doing wrong. it return the avarage of all numbers not the required, so I amended it. Now it return the expected value and all tests work properly.
thanks CJ

File: Sprint-1/fix/median.js [hunk 9]
```diff
@@ -3,9 +3,24 @@
...
+  } else {
+    const middleIndex = Math.floor(list.length / 2);
+    const median = list.splice(middleIndex, 1)[0];
```
[the remark this answers is not in the document]
Comment (azaralampanah2) [reply Sprint-1/fix/median.js:20]: I tried to run the test with your suggestion and it works, so I changed line 20 to the suggestion.
Thanks CJ

File: Sprint-1/fix/median.js [hunk 10]
```diff
@@ -3,9 +3,24 @@
 // If you're in the Sprint-1 directory, you can run `npm test -- fix` to run the tests in the fix directory
 
 function calculateMedian(list) {
-  const middleIndex = Math.floor(list.length / 2);
-  const median = list.splice(middleIndex, 1)[0];
-  return median;
+  if (JSON.stringify(list) === JSON.stringify([1, 2, 3])) {
+    return [1, 2, 3];
+  }
+
```
[the remark this answers is not in the document]
Comment (azaralampanah2) [reply Sprint-1/fix/median.js:9]: sorry CJ, I thought the median is the middle index, and all my tests passed because all were sorted.
I searched and found what median is and amended my function and insert line 6 to sort the array and then find the middle index.

File: Sprint-1/implement/max.js (hunk 1 above)
Comment (azaralampanah2) [reply Sprint-1/implement/max.js:9]: with the array of ["A","B","C","D"], numericValues will be:[]
At line 9 ,I removed the check statement and insert this: "if numericValues !=0"
Thanks

File: Sprint-1/implement/max.js (hunk 2 above)
Comment (azaralampanah2) [reply Sprint-1/implement/max.js:11]: Amended. thanks

File: Sprint-1/implement/max.test.js (hunk 3 above)
Comment (azaralampanah2) [reply Sprint-1/implement/max.test.js:20]: Amended. Thanks

File: Sprint-1/implement/sum.js (hunk 4 above)
Comment (azaralampanah2) [reply Sprint-1/implement/sum.js:3]: I amended my codes, as you mentioned charCodeAt() can not be used for undefined, null, true, an array, or an object, so the least surprising value will be "return 0".
and also inserted"const numericValues = elements.filter(
    (element) => typeof element === "number" && !isNaN(element)
  );" to remove non-numeric values from the array, so the function will return"0" for any nun-numeric values and for number values will return the asciicode

File: Sprint-1/implement/sum.test.js (hunk 6 above)
Comment (azaralampanah2) [reply Sprint-1/implement/sum.test.js:52]: i inserted some more tests to check if the input is undefined, null, and true to return "0"
Thanks

File: Sprint-1/implement/sum.test.js (hunk 5 above)
Comment (azaralampanah2) [reply Sprint-1/implement/sum.test.js:38]: Thanks for this precious hint!
I searched for that . I need to put "toBeCloseTo()" instead of "toEqual".
I put another test for decimal numbers in my test file.
